### DISCUSSION TITLE
ODBClient - element_connectivity method is quite time consuming for larger .odb files

### DISCUSSION THREAD

**Discussion by KarasTomas** [boschresearch/pylife#120] [General]
Hello,

We are writing script to extract results from Abaqus .odb into .vmap (or HDF5) file using pylife and its `odbClient`. We have an issue with `element_connectivity` method. It is quite time consuming. It takes about 160 s to load and save the data of approximately 50 k nodes mesh. 

The 160 s is length of whole process from start to save (geometry of one instance and stress variable "S" in one frame of one step), where `element_connectivity` method took around 150 s.  

<img width="562" alt="element_connectivity_test" src="https://github.com/user-attachments/assets/64659728-eb1d-4bc8-98e5-5cb5b83ca503">

We are using the `element_connectivity` method to construct a two level MultiIndex Pandas DataFrame (same as in Mesh Class) necessary for other applications. Namely saving the instance geometry to vmap or saving stress variable after averaging values at nodes.

[element_connectivity_test.zip](https://github.com/user-attachments/files/17544987/element_connectivity_test.zip)
I've included the code used in attachment and I can provide the .odb file used (but it is 0.8 GB). Uploaded code is modified to use .odb file from /demos/kt1.odb (which is significantly smaller)

I would assume that the long load time is caused by the connectivity DataFrame construction as it has list of nodes for each element with variable/not fixed number of nodes. That is why I includes different approach where i get the connectivity (or rather mesh) from the stress variable "S". But this is not ideal for cases stress might not be calculated. This speeds up the process form 160 s to about 15 s.

**Do you have any ideas where could be the problem and how could be `element_connectivity` speeded up?**

Thank you in advance for any advice, kind regards
Tomas
CTU in Prague

**Comment by johannes-mueller** [comment]
Hi @KarasTomas 

thank you very much for reaching out. I have done some profiling and performance tests. This indeed seems to be an issue that needs to be solved. At this point I cannot tell how quickly we can get it solved. 

You can track the progress in #121, closing the discussion here.

**Comment by johannes-mueller** [comment]
@KarasTomas Please try again with the new pyLife-2.1.3 release. Note that you also have to update `odbserver` in its environment.

**Comment by KarasTomas** [comment]
@johannes-mueller 
Thank you for dealing with this so quickly. I'll try the new release and let you know.

Edit: 
The issue seems to be fixed. The cell shown in the beginning of this conversation with previous `element_connectivity` load time of 150 s was reduced to 3.4 s. 

Thank you again for resolving this.
Tomas
